crates/fluvio-spu/src/replication/leader/kv.rs

```diff
@@ -0,0 +1,367 @@
+use std::{sync::Arc, task::Poll, pin::Pin};
+
+use anyhow::Result;
+use fluvio_kv_storage::{Log, LogBasedKVStorage};
+use fluvio_protocol::{
+    record::{RecordSet, RawRecords, Batch, Record},
+    Version,
+};
+use fluvio_storage::{
+    ReplicaStorage,
+    iterators::{FileBatchIterator, FileRecordIterator, RecordItem},
+};
+use futures_util::{Stream, ready, Future, FutureExt};
+use tracing::debug;
+
+use super::{LeaderReplicaState, FollowerNotifier};
+
+const RECORDS_SERIALIZATION_VERSION: Version = 0;
+
+pub type LeaderKVStorage<K, V, S> = LogBasedKVStorage<K, V, LeaderReplicaLog<S>>;
+
+/// [`Log`] implementation that enables using [`fluvio_kv_storage::KVStorage`] on top of [`LeaderReplicaState`].
+/// Basically, it implements a logic for storing KV operations on the topic, and reading them from
+/// the end of the topic.
+#[derive(Debug)]
+pub struct LeaderReplicaLog<S> {
+    replica: LeaderReplicaState<S>,
+    follower_notifier: Arc<FollowerNotifier>,
+}
+
+impl<S> LeaderReplicaLog<S> {
+    pub fn new(replica: LeaderReplicaState<S>, follower_notifier: Arc<FollowerNotifier>) -> Self {
+        Self {
+            replica,
+            follower_notifier,
+        }
+    }
+}
+
+impl<S: ReplicaStorage + Send + Sync + 'static> Log for LeaderReplicaLog<S> {
+    async fn read_from_end(&self) -> Result<impl futures_util::Stream<Item = Result<Vec<u8>>>> {
+        Ok(ReplicaRevStream::new(self.replica.clone()))
+    }
+
+    async fn append_batch(&mut self, entries: Vec<Vec<u8>>) -> Result<()> {
```

Comment:
Can we crate some abstraction for for key value which I assume this is for.   Just adding bytes seems to be bit raw

Reply:
I think we have enough abstractions here. Not sure I understand your questions.

Reply:
Try to add more description which hope to clarify purpose